File: backend/app/Repositories/NotificationRepository.php
<?php

namespace App\Repositories;

use App\Models\Notification;

class NotificationRepository
{
    public function create(array $data): Notification
    {
        return Notification::create($data);
    }

    public function getUnread(int $userId)
    {
        return Notification::where('user_id', $userId)
            ->where('is_read', false)
            ->latest()
            ->take(10)
            ->get();
    }

    public function paginate(int $userId, int $perPage = 20)
    {
        return Notification::where('user_id', $userId)
            ->latest()
            ->paginate($perPage);
    }

    public function markRead(int $id, int $userId): Notification
    {
        $notification = Notification::where('id', $id)
            ->where('user_id', $userId)
            ->firstOrFail();
        $notification->update(['is_read' => true]);
        return $notification;
    }

    public function markAllRead(int $userId): int
    {
        return Notification::where('user_id', $userId)
            ->where('is_read', false)
            ->update(['is_read' => true]);
    }

    public function unreadCount(int $userId): int
    {
        return Notification::where('user_id', $userId)
            ->where('is_read', false)
            ->count();
    }
}

File: backend-dotnet/Repositories/NotificationRepository.cs
using Microsoft.EntityFrameworkCore;
using TexTrack.Api.Data;
using TexTrack.Api.Infrastructure;
using TexTrack.Api.Models;

namespace TexTrack.Api.Repositories;

public class NotificationRepository
{
    private readonly AppDbContext _db;
    public NotificationRepository(AppDbContext db) => _db = db;

    public async Task<Notification> CreateAsync(Notification notification)
    {
        _db.Notifications.Add(notification);
        await _db.SaveChangesAsync();
        return notification;
    }

    public Task<PagedResult<Notification>> PaginateAsync(long userId, int perPage = 20)
    {
        var query = _db.Notifications
            .Where(n => n.UserId == userId)
            .OrderByDescending(n => n.Id);
        var filters = new Dictionary<string, string?> { ["per_page"] = perPage.ToString() };
        return query.PaginateAsync(filters, perPage);
    }

    public async Task<Notification> MarkReadAsync(long id, long userId)
    {
        var notification = await _db.Notifications
            .FirstOrDefaultAsync(n => n.Id == id && n.UserId == userId)
            ?? throw new NotFoundException("Notification not found");
        notification.IsRead = true;
        await _db.SaveChangesAsync();
        return notification;
    }

    public async Task<int> MarkAllReadAsync(long userId)
    {
        var unread = await _db.Notifications
            .Where(n => n.UserId == userId && !n.IsRead)
            .ToListAsync();
        foreach (var n in unread) n.IsRead = true;
        await _db.SaveChangesAsync();
        return unread.Count;
    }

    public Task<int> UnreadCountAsync(long userId) =>
        _db.Notifications.CountAsync(n => n.UserId == userId && !n.IsRead);
}
